### PR TITLE
Further reworking of event iteration interface to try to fix #2312

### LINKED ISSUE #2312: event with variables in their name are handled incorrectly.
wesnoth hndles events with variales in their names incorrectl in multiple ways:

1) The events are currently not executed in the order as they appear in the wml, (events without variables are executed first then event with variables are executed)

2) the namecheck now happens eariler than before, which is a behviour change in case that one event changes a variable then it then used in the evenname of another event.

3) THe space ->underscrore replacement breaks eventnamees, for eventnames with spaces in their inline formula (`name = "turn $($num_turns  - 1)"`)  or an even easier  example `turn $num_turns end"` which would become `turn_$num_turns_end"`

### PR DESCRIPTION
**Pull request by Vultraz**
Turns out constructing a list of matching handlers breaks some undocumented behavior, namely events being executed in the order they're added (usually the order in which they appear in the WML). It also broke certain variable substitution by moving variable interpolation wholly before any event execution, as opposed to between each event.

This commit is an experiment with an even simpler method: simply iterating through every registered handler in order, checking for a name match on each one, and if it passes, immediately executing the provided handler.

Not ready yet, still needs some polishing.

### CODE REVIEW

**Comment by Vultraz** [comment]
Yay! Tests pass!

**Comment by AI0867** [comment]
Is it possible that an exception could be thrown between the recursion counter increment and decrement? In that case, a RAII-based counter would be better.

File: src/game_events/manager.cpp [hunk 1]
```diff
@@ -119,26 +121,65 @@ void manager::write_events(config& cfg) const
...
+
+			if(utils::might_contain_variables(name)) {
+				// If we don't have gamedata, we can't interpolate variables, so there's
+				// no way the name will match. Move on to the next one in that case.
+				if(!gd) {
+					continue;
+				}
+
+				matches = standardized_event_id ==
+					event_handlers::standardize_name(utils::interpolate_variables_into_string(name, *gd));
```
Comment (CelticMinstrel) [inline src/game_events/manager.cpp]: I think a `standardize_name` function is actually slightly _less_ efficient than a comparison that treats space and underscores as equal.

File: src/game_events/manager.cpp [hunk 2]
```diff
@@ -119,26 +121,65 @@ void manager::write_events(config& cfg) const
...
+			continue;
+		}
+
+		// Could be more than one.
+		for(const std::string& name : handler->names()) {
+			bool matches = false;
+
+			if(utils::might_contain_variables(name)) {
+				// If we don't have gamedata, we can't interpolate variables, so there's
+				// no way the name will match. Move on to the next one in that case.
```
Comment (CelticMinstrel) [inline src/game_events/manager.cpp]: This is _probably_ true, but if you wanted you could try to just _remove_ the variables (as if they were all empty). Not sure if that's a good idea though.

**Comment by Vultraz** [comment]
@AI0867 not entirely sure. What would the RAII method look like? Class with static counter that incremented in its ctor and decremented in the dtor?

**Comment by CelticMinstrel** [comment]
I can't see any obvious sources of potential exceptions, except maybe `interpolate_variables_into_string` but that probably catches the exception, right? (I'm thinking of if the string includes a formula with a syntax error, for example `$(.)`.)

**Comment by CelticMinstrel** [comment]
@Vultraz That's what RAII refers to, yes; you'd also need to indent everything between them one additional level too, since you need it to be decremented before that final if statement.

**Comment by AI0867** [comment]
@Vultraz Something along the lines of (it can probably be a bit less complicated):

    class event_cleaner {
        //...

        void decrement(){
            if(--counter == 0){
                event_handlers_->clean_up_expired_handlers(standardized_event_id);
            }
        }
        class recursion_entry{
            recursion_entry(event_cleaner* parent) : parent_(parent) {
                parent_->increment();
            }
            ~recursion_entry(){
                parent_->decrement();
            }
            //...
        };
        recursion_entry recurse(){
            return recursion_entry(this);
        }
    };

    void manager::execute_on_events(...){
        static auto ec = event_cleaner{event_handlers_};
        auto recursed = ec.recurse();
        //.. rest of the function
    }

**Comment by Vultraz** [comment]
@Pentarctagon Figured out why the new tests were hanging. `[end_turn]` only ends the current *side's* turn, not the turn for all sides.

**Comment by Pentarctagon** [comment]
Ah, my bad.  I wrote the WML in an existing scenario I have, and it only has 1 side.

**Comment by Vultraz** [comment]
Merging this now. I'll do some further cleanup and implement the RAII recursion counter later.

File: src/game_events/manager.cpp (hunk 1 above)
Comment (sapientN3T) [reply src/game_events/manager.cpp]: thanks, I’m glad someone else noticed that... unfortunately it looks like the old code is being tossed out despite being more efficient in the general case. The new code could only be more efficient in some contrived example with hundreds of commas— in other words, never.

Comment (Vultraz) [reply src/game_events/manager.cpp]: I figured the difference in performance should be minimal enough as to be unnoticeable, but I haven't profiled to see.

Comment (CelticMinstrel) [reply src/game_events/manager.cpp]: Well, you might be right, but the specialized comparison function has the additional benefit of not potentially breaking the event name when it contains a variable (if for whatever reason the variable hasn't yet been substituted). And in the worst case, the specialized comparison function runs through both strings once, while using the standardize_name runs through one of the strings twice. (An equality check runs through each string once; the standardize_name function runs through its argument once.) So, specialized comparison is 2n while standardize_name is 3n where n is the length of the string. Also, "worst-case" here is actually quite common, as it's the case where the names _are_ equal.

Comment (Vultraz) [reply src/game_events/manager.cpp]: I see. Perhaps I should restore the old comparison function.

**Pull request merged**